src/main.cpp

```diff
@@ -4225,41 +4227,62 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
+        // The way POS txs are structured is that the first output of first tx is empty with 0 value
         if (block.vtx[0].vout.size() != 1 || !block.vtx[0].vout[0].IsEmpty())
             return state.DoS(100, error("CheckBlock() : coinbase output not empty for proof-of-stake block"));
 
-        // Second transaction must be coinstake, the rest must not be
+        // Ensure there is only one stake in block and it's 1st tx:
         if (block.vtx.empty() || !block.vtx[1].IsCoinStake())
             return state.DoS(100, error("CheckBlock() : second tx is not coinstake"));
+
+        // Ensure there are no other stakes in tx
         for (unsigned int i = 2; i < block.vtx.size(); i++)
             if (block.vtx[i].IsCoinStake())
                 return state.DoS(100, error("CheckBlock() : more than one coinstake"));
 
-        // If consensus level checks are in place.
+        // After spork is enabled we'll perform additional POS checks
+        // If the spork is not enabled these conditions are NOT checked
         if (IsSporkActive(SPORK_23_STAKING_REQUIREMENTS) && block.GetBlockTime() >= GetSporkValue(SPORK_23_STAKING_REQUIREMENTS)) {
-            // Check for minimum value.
-            if (block.vtx[1].vout[1].nValue < Params().Stake_MinAmount())
-                return state.DoS(100, error("CheckBlock() : stake under min. stake value"));
+            CAmount minStakeAmount = Params().Stake_MinAmount();
+            unsigned int minStakeAge = nStakeMinAgeConsensus;
+            int minStakeConfirmations = Params().Stake_MinConfirmations();
 
-            // Check for coin age.
-            // First try finding the previous transaction in database.
+
+            // Get transaction of the staked input
             CTransaction txPrev;
             uint256 hashBlockPrev;
             if (!GetTransaction(block.vtx[1].vin[0].prevout.hash, txPrev, hashBlockPrev, true))
                 return state.DoS(100, error("CheckBlock() : stake failed to find vin transaction"));
-            // Find block in map.
+                
+            // Bulwark's "Re-Stake". Because Bulwark stores metadata identifying stake in tx we know that previously this was a POS reward
+            // If you have not previously staked on this input then you will have to wait longer for your stake to mature.
+            // This penalizes "Stake Grinding" and gives reason to leave stakes alone reducing traffic on the network.
+            if (IsSporkActive(SPORK_25_BWK_RESTAKE) && block.GetBlockTime() >= GetSporkValue(SPORK_25_BWK_RESTAKE)) {
+                // Time is doubled if it's not a basic steak (without split)
+                if (!txPrev.IsCoinStake() || txPrev.vout.size() != 3) {
+                    minStakeAge *= 2;
+                    minStakeConfirmations *= 2;
+                }
+            }
+            // Ensure the output of the stake is above min amount (100 for BWK)
+            if (block.vtx[1].vout[1].nValue < minStakeAmount)
```

Comment:
doesn't this achieve the same as the code deleted above?

Reply:
Yes the amount didn't change but I wanted to group the logic together.

Reply:
Ok